### ISSUE TITLE
Leads receive multiple mails during large campaign

### ISSUE THREAD

**Issue by wittwerch** [mautic/mautic#1737]
## Description

We encountered the problem, that during a large campaign some recipients received the same email more than once (up to 6 times!). Today we could locate the root cause: There were more than one process triggering the campaign running at the same time. The locking done in script_executions.json should ensure, that only one process is running at the same time. Unfortunately that mechanism is not working well for large dataset. Sending a large campaign can take up to 1.5 hour.
By default Mautic threats a command (like triggering a campaign) to be "dead" after a timeout of 1800s (30min). This means it will simply assume that the process is not running anymore and allow a new command to be launched.

app/bundles/CoreBundle/Command/ModeratedCommand.php

```
$timeout   = $this->getContainer()->hasParameter('mautic.command_timeout') ?
  $this->getContainer()->getParameter('mautic.command_timeout') : 1800;
..
if ($check + $timeout <= time()) {
  $this->executionTimes['in_progress'][$command][$key] = time();
} else {
```

In our case this leads to the situation, that approximately 3-4 process were running in parallel.
## Steps to reproduce
1. Create a mail campaign with 500 leads
2. Trigger the campaign (mautic:campaign:trigger)
3. Simulate a long running job by editing the script_executions.json. Just subtract 1800 from the unix timestamp.
4. Trigger the campaign again (mautic:campaign:trigger)

In our test around 720 mails were sent in that campaign => leads received the same email twice.
## Possible solution

I think we should get ride of the command_timeout and instead write the pid of the php process into the script_executions.json file (using a lock, see #1664). This would allow commands to run as long as they take. And still make sure, that the commands start again if the process died. You simply check if a process is there with the pid from the script_executions.json, otherwise the process died and you can start over. That would follow the same mechanism init-script are using on linux distributions.

If you agree with that change I would start to implement that improvement.

**Comment by jkuchar** [comment]
This makes sense for me.

However I'm not sure how this will work without `exec` enabled. (e.g. shared hosting) Any ideas?

**Comment by bamzi** [comment]
we too have found the same issue (though in our tests and can't go live because of it)

we're a bootstrapped startup (i.e. no money) and other solutions cost too much for the size of our list of ~150k and each of our campaigns send to ~10k people everyday.

**Comment by wittwerch** [comment]
> However I'm not sure how this will work without exec enabled. (e.g. shared hosting) Any ideas?

I think we can just check the proc system, so we don't need the exec command enabled.

```
if (file_exists( "/proc/$pid" )){
    //process with a pid = $pid is running
}
```

**Comment by escopecz** [comment]
Would that work on non-unix systems? Maybe Symfony's Process component could be used. Although I don't have any experience with it. http://symfony.com/doc/current/components/process.html

**Comment by jkuchar** [comment]
Example above will not work on Windows. And will not work on shared UNIX webservers due to [open_basedir restriction](http://php.net/manual/en/ini.core.php#ini.open-basedir).

**Comment by jkuchar** [comment]
Question is different - **do Mautic want to support these shared hostings?**

**Comment by escopecz** [comment]
That's a great question. The developers agree on getting rid of the web installer/updater and just use the Composer to deal with installations/updates. So CLI-based. It would solve the cache rebuilding issue on busy websites. But Mautic could not be installed on shared hostings because they usually don't support SSH. Then there are the normal people ( = opposite of developers) who say:

>  Let's not forget who is the average user of Mautic - a marketer

A marketer is usually scared of the command line interface. Right now it's not clear what the future versions will use. What do you guys think?

**Comment by jkuchar** [comment]
I think that it is very wrong if marketer is installing Mautic on server. There are too much infrastructure questions that he must (correctly) answer and set-up. I think that Mautic should be installed/updated by developers/dev-ops and used by marketers.

And to be honest I do not thing that support for shared hosting is must in these days. Docker is out so there are appearing hostings where you have full access into anything inside a container.

**Comment by alanhartless** [comment]
Majority of our users are on shared hosting so we can't just drop support for such a setup and require a VPS in order to install Mautic. Yes, some technical skill is required to get the cron jobs setup in such a way for best performance with the resources allotted to the hosted account. There has to be some realization that Mautic isn't going to spit out hundreds of thousands of emails in seconds from a shared host gator account. But I'm not convinced that means we should drop support for host gator shared hosting although I'd love to. Rather, we need to make Mautic able to better handle itself with these types of things. 

@wittwerch, you may want to look at https://github.com/mautic/mautic/issues/1664 which I think addresses this same issue (I haven't personally tested the code they proposed).

**Comment by alanhartless** [comment]
@wittwerch Just noticed you replied in that thread. So can I actually close this in preference to the other?

**Comment by wittwerch** [comment]
@alanhartless This issue and #1664 address to distinct problems, so please leave it open.

**Comment by nickion** [comment]
@witterch You suggested checking /proc. This may be useful, but it must be kept in mind that just because a given process ID may exist, it does not mean that it's a mautic related process. Using kill with signal 0 is another way to test for process existence as this effectively just performs validation on the PID. The potential problem is that the mautic process could have ended and a new process now exists with the same PID, and which by bad luck happens to be associated with a long running process. At that point mautic sending may be broken.

Another consideration is that even if the mautic process *is* still running, this does not mean that it's actually working. I'd be tempted to include a watchdog feature whereby the process needs to demonstrate that it is alive and actually doing something. An example might be to record the last time that it sent an email for a campaign. If there is no sign of activity, kill it if it exists and start a fresh process if needed,

**Comment by Salamek** [comment]
Hi guys, i HOTFIXed this problem by wrapping mautic processes into simple python script which is using its own lock files and checking if PID written in lock file is alive or not (i dont want to modify mautic src):

mautic_lock.py:
```python
import os
import sys
import subprocess

def check_pid(pid):        
    """ Check For the existence of a unix pid. """
    try:
        os.kill(pid, 0)
    except OSError:
        return False
    else:
        return True

action = sys.argv[1]

lock = os.path.join(os.path.dirname(os.path.abspath(__file__)), "mautic_{}.lock".format(action))

if os.path.isfile(lock):
    print("Lockfile {} exists".format(lock))
    with open(lock, 'r') as f:
        pid = f.read().strip()
        if check_pid(int(pid)):
           print("And PID {} is alive, exiting...".format(pid))
           sys.exit(0)
        else:
           print("And PID {} is dead, starting new proccess".format(pid))

proc = subprocess.Popen(['php', '/var/www/mautic.cvc.cz/www/app/console', 'mautic:{}'.format(action)], stdout=subprocess.PIPE)
with open(lock,"w") as f:
    f.write(str(proc.pid))
proc.wait()
os.remove(lock)
```

and in crontab i have:
```sh
# Original command
#*/15 * * * * php /var/www/mautic.cvc.cz/www/app/console mautic:campaigns:rebuild > /dev/null

# New command
*/15 * * * * python /root/mautic_lock.py campaigns:rebuild > /dev/null
#.... and so...
```

PS: This is only good for ppl with Linux OS and full access to system
PPS: I'm planning to implement database driven  "worker queue" into mautic later, current implementation is :hankey:

**Comment by wittwerch** [comment]
> You suggested checking /proc. This may be useful, but it must be kept in mind that just because a given process ID may exist, it does not mean that it's a mautic related process. Using kill with signal 0 is also another way to test for process existence as this effectively just performs validation on the PID. The potential problem however is that the mautic process could have ended and a new process now exists with the same PID, and which by bad luck happens to be associated with a long running process. At that point mautic sending may be broken.

PIDs will increase to limit (32768) and then wrap around. The kernel usually will not reuse a PID before this wrap-around happens. So your scenario is highly unlikely and you can still check the process name to make sure it is Mautic. I think relying on this approach is still way better than just assuming that a command won't run any longer than 30min.

In my opinion the whole cronjob setup is flawed. Having to wait to get segments updated by a cronjob is not state-of-the-art. I would prefer a system where you simple schedule such tasks in queues and then have worker processes, that pull work from these queues. A good example from the Python ecosystem is Celery (http://www.celeryproject.org/).
I think that's the way to go to build a reliable and scalable system, but that won't run on your 5$ hosting.

**Comment by jkuchar** [comment]
@wittwerch I also would like to see celery-like implementation of queues in Mautic. It will make processing much smoother, faster and more reliable. [RabbitMQ has nice driver for PHP](https://github.com/php-amqplib/php-amqplib). I've seen this in production on another project and it works perfectly!

With docker hosting you can have hypothetically running Mautic with RabitMQ for 7$/mo. Cool, right? https://cloud.docker.com/pricing/

**Comment by chandon** [comment]
There is already a system to lock command in symfony 
http://symfony.com/blog/new-in-symfony-2-6-lockhandler

**Comment by nickion** [comment]
@jkuchar Nice, and you're correct about server costs. I wish our spend were as little as $5 a month on servers, but these days one should be able to find a machine at a decent provider for $5 to $10 a month with sufficient relevant resources to handle the task of sending out a decent volume of emails without breaking a sweat.  Christian @wittwerch is right about failure being "highly unlikely", but note that not all Unixes issue PIDs in a round robin fashion, and many a system has had failures from adopting a "highly unlikely" to go wrong approach. It often doesn't take much to simply get it right in the first place, and why make a system that works predicated on assumptions that might well be wrong from the outset or become wrong in the future. There's another Mautic bug I reported a while back whereby a recipient reading an email would be counted as activity on a different lead. That issue might also be unlikely to happen but it happened reproducibly during due diligence testing, and would be particularly likely if multiple recipients had the same IP address. 

We're not talking about rocket science here, it's just a simple queue, and implementing a basic and rock solid queuing system is straightforward enough. The cronjob approach isn't flawed per-se, but it's not the only way and alternatives on balance may be better. Ultimately it's not so much about how it works but whether it does actually work. Provided that it works and most Mautic users will be able to use it, the implementation will be a detail that really does not matter.

**Comment by escopecz** [comment]
Fixed in 2.3
